bindings/python/pyiree/compiler2/tools.py

```diff
@@ -0,0 +1,244 @@
+
+
+def find_tool(exe_name: str) -> str:
+  """Finds a tool by its (extension-less) executable name.
+
+  Args:
+    exe_name: The name of the executable (extension-less).
+  Returns:
+    An absolute path to the tool.
+  Raises:
+    ValueError: If the tool is not known or not found.
+  """
+  if exe_name not in _TOOL_MODULE_MAP:
+    raise ValueError(f"IREE compiler tool '{exe_name}' is not a known tool")
+  # First search an explicit tool path.
+  tool_path = get_tool_path()
+  for path_entry in tool_path:
+    if not path_entry:
+      continue
+    candidate_exe = os.path.join(path_entry, exe_name)
+    if os.path.isfile(candidate_exe) and os.access(candidate_exe, os.X_OK):
+      return candidate_exe
+
+  # Attempt to load the tool module.
+  tool_module_name = _TOOL_MODULE_MAP[exe_name]
+  tool_module_package = _TOOL_MODULE_PACKAGES[tool_module_name]
+  try:
+    tool_module = importlib.import_module(tool_module_name)
+  except ModuleNotFoundError:
+    raise ValueError(
+        f"IREE compiler tool '{exe_name}' is not installed (it should have been "
+        f"found in the python module '{tool_module_name}', typically installed "
+        f"via the package {tool_module_package}).\n\n"
+        f"Either install the package or set the {_TOOL_PATH_ENVVAR} environment "
+        f"variable to contain the path of the tool executable "
+        f"(current {_TOOL_PATH_ENVVAR} = {repr(tool_path)})") from None
+
+  # Ask the module for its tool.
+  candidate_exe = tool_module.get_tool(exe_name)
+  if (not candidate_exe or not os.path.isfile(candidate_exe) or
+      not os.access(candidate_exe, os.X_OK)):
+    raise ValueError(
+        f"IREE compiler tool '{exe_name}' was located in module "
+        f"'{tool_module_name}' but the file was not found or not executable: "
+        f"{candidate_exe}")
+  return candidate_exe
+
+
+def invoke_immediate(command_line: List[str],
+                     *,
+                     input_file: Optional[str] = None,
+                     immediate_input=None):
+  """Invokes an immediate command.
+
+  This is separate from invoke_pipeline as it is simpler and supports more
+  complex input redirection, using recommended facilities for sub-processes
+  (less magic).
+  """
+  run_args = {}
+  input_file_handle = None
+  stderr_handle = sys.stderr
+  try:
+    # Redirect input.
+    if input_file is not None:
+      input_file_handle = open(input_file, "rb")
+      run_args["stdin"] = input_file_handle
+    elif immediate_input is not None:
+      run_args["input"] = immediate_input
+
+    # Capture output.
+    run_args["capture_output"] = True
+    process = subprocess.run(command_line, **run_args)
+    if process.returncode != 0:
+      raise CompilerToolError(process)
+    # Emit stderr contents.
+    _write_binary_stderr(stderr_handle, process.stderr)
+    return process.stdout
+  finally:
+    if input_file_handle:
+      input_file_handle.close()
+
+
+def invoke_pipeline(command_lines: List[List[str]]):
```

Comment:
It would be nice to have a comment on why using threading here is useful despite the pipeline being serial. 

I'm guessing it's because each individual process will execute what they can without `stdin` (e.g. starting up?) in parallel and then wait on the previous processes' stdin? I'm honestly not sure.

Reply:
I'll add a one liner, but in brief for education/explanation: the pipeline may look serial, but in fact all of the processes run in parallel and are coupled at the OS level by small (relative to typical input sizes) buffers. It is insanely easy to deadlock such pipelines by failing to service the buffers at any of the ends (which includes stderr in our case, since we capture it separately).

Native solutions will typically multiplex all of this with `select` and `wait` incantations without using threads. In Python, I've made myself very sad trying to do the same thing, and the API really isn't set up for it (at least without egregious, non-portable signal hacks). The typical way you use `subprocesses` is either via `run` (which handles it all for you) or `Popen(...).communicate(input)` which blocks and is good for piping in and out of one process (but doesn't give you fine grained control for constructing a pipeline). What I'm doing here is effectively what `communicate` does under the covers (firing up file handle pumping threads) -- it is just usually hidden from you.

At the scale that this will be running, a few extra threads is fine.